src/Orchard/Environment/DefaultOrchardHost.cs

```diff
@@ -343,6 +348,20 @@ public class DefaultOrchardHost : IOrchardHost, IShellSettingsManagerEventHandle
             _tenantsToRestart.GetState().Add(context.Settings);
         }
 
+        private void BlockRequestsDuringSetup() {
+            if (_shellContexts == null)
+                return;
+
+            // If there's only one tenant and it's initializing, return a Service Unavailable HTTP status code.

```

Comment:
Yeah I thought of that. If a request comes in for another tenant, then with this code it would just behave is before the Initializing status code changes (meaning clients would see the Setup screen even if the tenant is currently cooking).

To solve for that, I think we'd have to know what shell the current request matches. Once we know that, it should be simple to update this code to check for any shell's state rather than the first and only shell.
Can you think of any issues with such an implementation?